action/protocol/execution/evm/evmstatedbadapter_erigon.go

```diff
@@ -0,0 +1,227 @@
+package evm
+
+import (
+	"time"
+
+	erigonchain "github.com/erigontech/erigon-lib/chain"
+	libcommon "github.com/erigontech/erigon-lib/common"
+	types2 "github.com/erigontech/erigon-lib/types"
+	erigonstate "github.com/erigontech/erigon/core/state"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/holiman/uint256"
+
+	"github.com/iotexproject/go-pkgs/hash"
+
+	"github.com/iotexproject/iotex-core/v2/blockchain/genesis"
+	"github.com/iotexproject/iotex-core/v2/pkg/log"
+	"github.com/iotexproject/iotex-core/v2/state"
+)
+
+// ErigonStateDBAdapter is an adapter for persisted both in erigon intra block state and statedb
+// reads are based on statedb
+type ErigonStateDBAdapter struct {
+	*StateDBAdapter
+	intra *erigonstate.IntraBlockState
+}
+
+// ErigonStateDBAdapterDryrun is an adapter for simulation
+// reads are based on erigon intra block state, writes not persisted
+type ErigonStateDBAdapterDryrun struct {
+	*ErigonStateDBAdapter
+}
+
+// NewErigonStateDBAdapter creates a new ErigonStateDBAdapter
+func NewErigonStateDBAdapter(adapter *StateDBAdapter,
+	intra *erigonstate.IntraBlockState,
+) *ErigonStateDBAdapter {
+	adapter.newContract = func(addr hash.Hash160, account *state.Account) (Contract, error) {
+		return newContractAdapter(addr, account, adapter.sm, intra, adapter.asyncContractTrie)
+	}
+	return &ErigonStateDBAdapter{
+		StateDBAdapter: adapter,
+		intra:          intra,
+	}
+}
+
+// NewErigonStateDBAdapterDryrun creates a new ErigonStateDBAdapterDryrun
+func NewErigonStateDBAdapterDryrun(adapter *StateDBAdapter,
+	intra *erigonstate.IntraBlockState,
+) *ErigonStateDBAdapterDryrun {
+	a := NewErigonStateDBAdapter(adapter, intra)
+	adapter.newContract = func(addr hash.Hash160, account *state.Account) (Contract, error) {
+		return newContractErigon(addr, account, intra)
+	}
+	return &ErigonStateDBAdapterDryrun{
+		ErigonStateDBAdapter: a,
+	}
+}
+
+// SubBalance subtracts the balance of the given address by the given value
+func (s *ErigonStateDBAdapter) SubBalance(evmAddr common.Address, v *uint256.Int) {
+	s.StateDBAdapter.SubBalance(evmAddr, v)
+	// balance updates for erigon will be done in statedb
+}
+
+// AddBalance adds the balance of the given address by the given value
+func (s *ErigonStateDBAdapter) AddBalance(evmAddr common.Address, v *uint256.Int) {
+	s.StateDBAdapter.AddBalance(evmAddr, v)
+	// balance updates for erigon will be done in statedb
+}
+
+// SetCode set the code of the given address
+func (s *ErigonStateDBAdapter) SetCode(evmAddr common.Address, c []byte) {
+	s.StateDBAdapter.SetCode(evmAddr, c)
+	// code updates for erigon will be done by Contract
```

Comment:
why here defer to Contract to update for erigon, and some methods below (like SetNonce(), AddRefund()) erigon did by itself